_posts/2021-11-19-connecting-java-high-level-rest-client-with-opensearch-over-https.markdown

```diff
@@ -0,0 +1,150 @@
+
+Let’s start with including the client dependency in `build.gradle`. This will download the dependency from maven central when the project is executed. It is recommended to use the same version of client as OpenSearch version.
+
+```
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    implementation 'org.opensearch.client:opensearch-rest-high-level-client:1.1.0'
+}
+```
+
+Next, we will create an instance of `RestHighLevelClient` and use that to create an index and ingest some data into OpenSearch. But before going there, hold on a sec! Remember, while setting up the server we configured SSL certificates which enabled HTTPS (and disabled HTTP) on server side? We’ll need to configure the client as well to successfully establish an SSL connection with the server. Let’s see how to configure the client truststore.
+
+A Java application (by default) uses the JVM truststore, which holds certificates from the trusted [Certified Authorities (CA)](https://en.wikipedia.org/wiki/Certificate_authority), to verify the certificate presented by the server in an SSL connection. You can use the Java `keytool` to see the list of trusted CAs in your JVM truststore.
+
+```
+keytool -keystore $JAVA_HOME/lib/security/cacerts -storepass changeit -list 
+```
+
+To configure the client, we need to add the root CA certificate (root-ca.pem) to the client truststore. The install_demo_configurations.sh tool created `opensearch-1.1.0/config/root-ca.pem` while setting up the server. We can either add it to the JVM truststore, or add it to a custom truststore and use that custom truststore in the client. I prefer the custom truststore approach to keep the JVM truststore clean. 
+
+Use the Java `keytool` to create a custom truststore and import the root authority certificate. The `keytool` does not understand the `.pem` format, so we’ll have to first convert the root authority certificate to `.der` format using `openssl` cryptography library and then add it to the custom truststore using Java `keytool`. Most Linux distributions already come with `openssl` installed.
+
+_STEP 1:_ Convert the root authority certificates from .pem to .der format.
+
+```
+openssl x509 -in opensearch-1.1.0/config/root-ca.pem -inform pem -out root-ca.der --outform der
+```
+
+_STEP 2:_ Create a custom truststore and add the `root-ca.der` certs. Later, we will set this custom truststore in the client code to override the default truststore.
+
+```
+keytool -import root-ca.der -alias opensearch -keystore myTrustStore
+```
+
+Confirm the action was successful by listing certs in truststore. The grep should be able to find opensearch alias if the certs were added successfully.
+
+```
+keytool -keystore myTrustStore -storepass changeit -list | grep opensearch
+```
+
+_STEP 3:_ Next, set the truststore properties in the client code to point to the custom truststore. 
```

Comment:
'in the client code' confuses me. Is this the code of the high level rest client or is this the code of the example? I would tighten up the language a bit here to be more specific

Reply:
It refers to the code where client is being used. Does this make more sense? - `Next, set the truststore properties in your Java application code to point to the custom truststore`